### PR TITLE
feature/manage-sell

### PR DESCRIPTION
**Pull request by Le-Caignec**
Smart Contract : 
- Develop Sale.sol contract ✅
- add removeProtectedDataFromSubscription function ✅
- add modifier for subscription : onlyCollectionNotSubscribed & onlyProtectedDataNotAvailableInSubscription ✅
- add modifier for renting : onlyProtectedDataNotRented & onlyProtectedDataNotForRenting ✅
- refactor wording tenants to renters ✅

Tests : 
- add corresponding tests for Sale contract ✅
- modify & add tests for Subscription contract with the new modifiers ✅

### CODE REVIEW

File: packages/sharing-smart-contract/contracts/modules/Sale.sol [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+     ***************************************************************************/
+    function setProtectedDataForSale(
+        uint256 _collectionId,
+        address _protectedData,
+        uint112 _price
+    ) public onlyProtectedDataInCollection(_collectionId, _protectedData) {
+        protectedDataForSale[_collectionId][_protectedData].onSale = true;
+        protectedDataForSale[_collectionId][_protectedData].price = _price;
+        emit ProtectedDataAddedForSale(_collectionId, _protectedData, _price);
+    }
```
Comment (PierreJeanjacquot) [inline packages/sharing-smart-contract/contracts/modules/Sale.sol:45]: we should check
- the data is not available for renting or subscription
- there is no pending rental or subscription on the data

File: packages/sharing-smart-contract/contracts/modules/Collection.sol [hunk 2]
```diff
@@ -97,4 +101,15 @@ contract Collection is ERC721Burnable, ERC721Receiver {
...
+    function swapCollection(
+        uint256 _collectionIdFrom,
+        uint256 _collectionIdTo,
+        address _protectedData
+    ) external onlyOwner {
+        delete protectedDatas[_collectionIdFrom][uint160(_protectedData)];
+        emit RemoveProtectedDataFromCollection(_collectionIdFrom, _protectedData);
+        protectedDatas[_collectionIdTo][uint160(_protectedData)] = _protectedData;
+        emit AddProtectedDataToCollection(_collectionIdTo, _protectedData);
+    }
```
[the remark this answers is not in the document]
Comment (Le-Caignec) [reply packages/sharing-smart-contract/contracts/modules/Collection.sol]: enable ProtectedDataSharing contract to change protectedData from one mapping to an other one
Commit: 7a9763727a83350c7ce4a38bdfc2ae17375db535

File: packages/sharing-smart-contract/contracts/Store.sol [hunk 3]
```diff
@@ -78,4 +78,15 @@ abstract contract Store {
...
+    /***************************************************************************
+     *                       Saling                                            *
+     ***************************************************************************/
+    // collectionId => (ProtectedDataTokenId => RentingParams)
+    mapping(uint256 => mapping(address => SaleingParams)) public protectedDataForSale;
+
+    struct SaleingParams {
+        bool inSaling;
+        uint112 price; // 112 bit allows for 10^15 eth
+    }
```
[the remark this answers is not in the document]
Comment (Le-Caignec) [reply packages/sharing-smart-contract/contracts/Store.sol]: Commit: 66e7f12cface01ce28e4061e84df88a87a295fde
Commit: a6eafec43bac33c40ebdd14f52af931da920421d

File: packages/sharing-smart-contract/test/sale.test.js [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+      protectedDataSharingContract,
+      collectionContract,
+      collectionTokenIdFrom,
+      collectionTokenIdTo,
+      addr1,
+      addr2,
+    };
+  }
+
+  async function addProtectedDataToCollectionHelper(collectionContract, collectionTokenId, addr) {
```
[the remark this answers is not in the document]
Comment (Le-Caignec) [reply packages/sharing-smart-contract/test/sale.test.js]: Commit: 053e4e3e9211b6e19bd93e9b03ba9b3b4cdb65e2

File: packages/sharing-smart-contract/contracts/modules/Sale.sol [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+            "Wrong amount sent"
+        );
+        m_collection.swapCollection(
+            _collectionIdFrom,
+            _collectionIdTo,
+            _protectedData
+        );
+        delete protectedDataForSale[_collectionIdFrom][_protectedData];
+        emit ProtectedDataSold(_collectionIdFrom, _collectionIdTo, _protectedData);
+    }
```
[the remark this answers is not in the document]
Comment (Le-Caignec) [reply packages/sharing-smart-contract/contracts/modules/Sale.sol]: add two different interfaces

File: packages/sharing-smart-contract/contracts/Store.sol [hunk 6]
```diff
@@ -59,23 +59,79 @@ abstract contract Store {
...
 
     struct SubscriptionParams {
         uint112 price; // 112 bit allows for 10^15 eth
         uint48 duration; // 48 bit allows 89194 years of delay
     }
 
+    modifier onlyCollectionNotSubscribed(uint256 _collectionId) {
+        require(
+            lastSubscriptionExpiration[_collectionId] < block.timestamp,
+            "Collection subscribed"
```
Comment (PierreJeanjacquot) [inline packages/sharing-smart-contract/contracts/Store.sol:75]: A bit more explicit
```suggestion
            "Collection has ongoing subscriptions"
```

File: packages/sharing-smart-contract/contracts/Store.sol [hunk 7]
```diff
@@ -59,23 +59,79 @@ abstract contract Store {
     mapping(uint256 => mapping(address => bool)) public protectedDataInSubscription;
     // collectionId => (subscriberAddress => endTimestamp(48 bit for full timestamp))
     mapping(uint256 => mapping(address => uint48)) public subscribers;
-    //contentCreatorId => subscriber
+    // collectionId => subscriber
```
Comment (PierreJeanjacquot) [inline packages/sharing-smart-contract/contracts/Store.sol:62]: oudated comment
```suggestion
    // collectionId => subscriptionParams:  SubscriptionParams
```

File: packages/sharing-smart-contract/contracts/Store.sol [hunk 8]
```diff
@@ -59,23 +59,79 @@ abstract contract Store {
...
+        _;
+    }
+
+    modifier onlyProtectedDataNotAvailableInSubscription(
+        uint256 _collectionId,
+        address _protectedData
+    ) {
+        require(
+            protectedDataInSubscription[_collectionId][_protectedData] == false,
+            "ProtectedData available for subscription"
```
Comment (PierreJeanjacquot) [inline packages/sharing-smart-contract/contracts/Store.sol:86]:
```suggestion
            "ProtectedData is available in subscription"
```

File: packages/sharing-smart-contract/contracts/Store.sol [hunk 9]
```diff
@@ -59,23 +59,79 @@ abstract contract Store {
...
+        );
+        _;
+    }
     /***************************************************************************
      *                       Renting                                      *
      ***************************************************************************/
     // collectionId => (ProtectedDataTokenId => RentingParams)
```
Comment (PierreJeanjacquot) [inline packages/sharing-smart-contract/contracts/Store.sol:93]: outdated comment
```suggestion
    // collectionId => (protectedDataAddress: address => rentingParams: RentingParams)
```

File: packages/sharing-smart-contract/contracts/Store.sol [hunk 10]
```diff
@@ -59,23 +59,79 @@ abstract contract Store {
...
+    mapping(address => uint48) public lastRentalExpiration;
 
     struct RentingParams {
         bool inRenting;
         uint112 price; // 112 bit allows for 10^15 eth
         uint48 duration; // 48 bit allows 89194 years of delay
     }
+
+    modifier onlyProtectedDataNotRented(address _protectedData) {
+        require(lastRentalExpiration[_protectedData] < block.timestamp, "ProtectedData rented");
```
Comment (PierreJeanjacquot) [inline packages/sharing-smart-contract/contracts/Store.sol:107]: not sure
```suggestion
        require(lastRentalExpiration[_protectedData] < block.timestamp, "ProtectedData is currently being rented");
```

File: packages/sharing-smart-contract/contracts/Store.sol [hunk 11]
```diff
@@ -59,23 +59,79 @@ abstract contract Store {
...
+            protectedDataForRenting[_collectionId][_protectedData].inRenting == false,
+            "ProtectedData available for renting"
+        );
+        _;
+    }
+
+    /***************************************************************************
+     *                       Saling                                            *
+     ***************************************************************************/
+    // collectionId => (ProtectedDataTokenId => SellingParam)
```
Comment (PierreJeanjacquot) [inline packages/sharing-smart-contract/contracts/Store.sol:122]: outdated comment
```suggestion
    // collectionId => (protectedDataAddress: address => sellingParam: SellingParam)
```

File: packages/sharing-smart-contract/contracts/modules/Collection.sol
Comment (PierreJeanjacquot) [inline packages/sharing-smart-contract/contracts/modules/Collection.sol:1]: should be changed in the future (out of scope)

File: packages/sharing-smart-contract/contracts/modules/Sale.sol [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+    ) public payable onlyCollectionOwner(_collectionIdTo) {
+        require(
+            protectedDataForSale[_collectionIdFrom][_protectedData].forSale,
+            "ProtectedData not for sale"
+        );
+        require(
+            protectedDataForSale[_collectionIdFrom][_protectedData].price == msg.value,
+            "Wrong amount sent"
+        );
+        m_collection.adminSwapCollection(_collectionIdFrom, _collectionIdTo, _protectedData);
```
Comment (PierreJeanjacquot) [inline packages/sharing-smart-contract/contracts/modules/Sale.sol:70]: should change in the future (out of scope)

File: packages/sharing-smart-contract/contracts/modules/Sale.sol [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+        require(
+            protectedDataForSale[_collectionIdFrom][_protectedData].forSale,
+            "ProtectedData not for sale"
+        );
+        require(
+            protectedDataForSale[_collectionIdFrom][_protectedData].price == msg.value,
+            "Wrong amount sent"
+        );
+        delete protectedDataForSale[_collectionIdFrom][_protectedData];
+        m_collection.adminSafeTransferFrom(_to, _protectedData);
```
Comment (PierreJeanjacquot) [inline packages/sharing-smart-contract/contracts/modules/Sale.sol:89]: should change in the future (out of scope)

File: packages/sharing-smart-contract/contracts/modules/Subscription.sol
Comment (PierreJeanjacquot) [inline packages/sharing-smart-contract/contracts/modules/Subscription.sol:1]: missing modifier `onlyProtectedDataNotForSale(uint256 _collectionId, address _protectedData)` on `setProtectedDataToSubscription`

File: packages/sharing-smart-contract/contracts/modules/Renting.sol [hunk 14]
```diff
@@ -30,27 +30,48 @@ contract Renting is Store {
...
+        }
+        emit NewRental(_collectionId, _protectedData, endDate);
+    }
+
     function setProtectedDataToRenting(
         uint256 _collectionId,
         address _protectedData,
         uint112 _price,
         uint48 _duration
     ) public onlyProtectedDataInCollection(_collectionId, _protectedData) {
```
Comment (PierreJeanjacquot) [inline packages/sharing-smart-contract/contracts/modules/Renting.sol:61]: Missing modifier `onlyProtectedDataNotForSale(uint256 _collectionId, address _protectedData)`

File: packages/sharing-smart-contract/test/subscription.test.js [hunk 15]
```diff
@@ -172,25 +239,12 @@ describe('Subscription.sol', () => {
 
   describe('setProtectedDataToSubscription()', () => {
```
Comment (PierreJeanjacquot) [inline packages/sharing-smart-contract/test/subscription.test.js:240]: add test case
- cannot setProtectedDataToSubscription if the protected data is available for sale

File: packages/sharing-smart-contract/test/subscription.test.js [hunk 16]
```diff
@@ -215,4 +269,60 @@ describe('Subscription.sol', () => {
       ).to.be.revertedWith('ProtectedData is not in collection');
     });
   });
+
+  describe('removeProtectedDataFromSubscription()', () => {
```
Comment (PierreJeanjacquot) [inline packages/sharing-smart-contract/test/subscription.test.js:273]: add test case
- cannot removeProtectedDataFromSubscription if the collection has ongoing subscriptions

File: packages/sharing-smart-contract/test/renting.test.js
Comment (PierreJeanjacquot) [inline packages/sharing-smart-contract/test/renting.test.js:1]: add test case in `setProtectedDataToRenting()`
- cannot setProtectedDataToRenting if the protected data is available for sale

**Review by PierreJeanjacquot — CHANGES_REQUESTED** [review]

File: packages/sharing-smart-contract/contracts/Store.sol [hunk 17]
```diff
@@ -59,23 +59,82 @@ abstract contract Store {
...
+    modifier onlyProtectedDataNotForRenting(uint256 _collectionId, address _protectedData) {
+        require(
+            protectedDataForRenting[_collectionId][_protectedData].inRenting == false,
+            "ProtectedData available for renting"
+        );
+        _;
+    }
+
+    /***************************************************************************
+     *                       Saling                                            *
```
Comment (PierreJeanjacquot) [inline packages/sharing-smart-contract/contracts/Store.sol:123]:
```suggestion
     *                       Selling                                           *
```

File: packages/sharing-smart-contract/contracts/modules/Renting.sol [hunk 18]
```diff
@@ -30,27 +30,52 @@ contract Renting is Store {
...
+        onlyProtectedDataNotForSale(_collectionId, _protectedData)
+    {
+        require(_duration > 0, "Duration param invalide");
+        protectedDataForRenting[_collectionId][_protectedData].inRenting = true;
+        protectedDataForRenting[_collectionId][_protectedData].price = _price;
+        protectedDataForRenting[_collectionId][_protectedData].duration = _duration;
         emit ProtectedDataAddedToRenting(_collectionId, _protectedData, _price, _duration);
     }
 
+    // cannot be rented anymore, pending rental are still valid
```
Comment (PierreJeanjacquot) [inline packages/sharing-smart-contract/contracts/modules/Renting.sol:73]:
```suggestion
    // cannot be rented anymore, ongoing rental are still valid
```

**Review by PierreJeanjacquot — APPROVED** [review]

**Pull request merged**